### DISCUSSION TITLE
Adding support for JPEG XL as the new default

### DISCUSSION THREAD

**Discussion by mackoj** [pointfreeco/swift-snapshot-testing#898] [General]
Hi,

What do you think about migrating the default PNG format to JXL format for this lib ?

JPEG XL(`.jxl`) has been inside macOS and iOS since last year and it's better that HEIC. I do think it could be a great new default for `swift-snapshot-testing` and we could provide a fallback for platforms that did not have it using the [libjxl](https://github.com/libjxl/libjxl).

In my test my big snapshot when from 326.3mb to 4.1mb(**98.74% size decrease**) in average in a folder of 1000 files it size decrease was more around 70%.

<img width="935" alt="Screenshot 2024-08-31 at 2 52 18 PM" src="https://github.com/user-attachments/assets/2d669225-0141-4bbd-8be6-ddcfab41ca54">

## What is JPEG XL

Using JPEG XL is advantageous in several scenarios, especially as its adoption grows. Here are some key reasons to consider using JPEG XL:

### 1. **Efficient Compression**
   - **Smaller File Sizes**: JPEG XL offers both lossy and lossless compression with significantly smaller file sizes compared to older formats like PNG and JPEG. This efficiency is particularly beneficial for web use, reducing bandwidth and improving load times.

### 2. **High Image Quality**
   - **Preserves Detail**: JPEG XL maintains excellent image quality, even at lower bitrates, making it ideal for high-resolution photos and graphics where quality is crucial.

### 3. **Support for Transparency**
   - **Alpha Channels**: Unlike traditional JPEG, JPEG XL supports transparency, which is essential for modern web and graphic design, allowing for images with varying levels of opacity.

### 4. **Future-Proofing**
   - **HDR and Wide Color Gamut**: JPEG XL supports HDR (High Dynamic Range) and extended color gamuts, making it suitable for future high-end displays and photographic needs.

### 5. **Backward Compatibility**
   - **Seamless Transition**: JPEG XL can embed original JPEG data, allowing for a smooth transition from older JPEG formats without losing compatibility or quality.

### 6. **Faster Performance**
   - **Quick Encoding/Decoding**: JPEG XL is designed for faster encoding and decoding, which is useful in performance-critical environments like web applications and mobile devices.

### 7. **Versatility**
   - **One Format for All**: JPEG XL can replace multiple image formats (JPEG, PNG, GIF) thanks to its versatility, reducing the need to use different formats for different purposes.

### **JPEG XL Features**
- Smaller files means faster loading times and reduced bandwidth usage.
- Maintain high quality while saving storage space.
- When working with images that require transparency and high color fidelity.
- For storing high-quality images with the option of lossless compression.
- When preparing for HDR content and displays with wide color gamuts.

JPEG XL is a strong choice when you need efficient compression, high quality, and modern features like transparency and HDR support, making it suitable for a wide range of applications.

**Comment by mbrandonw** [comment]
Hi @mackoj, thanks for bring up this topic, but since it is not an issue with the library I am going to convert it to a discussion.

**Comment by mackoj** [comment]
A prototype has been done here https://github.com/pointfreeco/swift-snapshot-testing/pull/899.

**Comment by mackoj** [comment]
The PR above adds support for `.jxl`, `.heic`, and `.webp` formats. The goal is to test which format performs best and to provide users with options to choose according to their needs. It's important to note that this change will not affect existing users of the library because the new formats must be explicitly used and linked in the test target to take effect.
